chained/water_impl.go

```diff
@@ -115,11 +137,45 @@ func createDialer(ctx context.Context, wasm []byte, transport string) (water.Dia
 	return dialer, nil
 }
 
+func (d *waterImpl) isReady() (bool, error) {
+	d.readyMutex.Lock()
+	defer d.readyMutex.Unlock()
+	return d.ready, d.errLoadingWASM
+}
+
+func (d *waterImpl) setReady() {
+	d.readyMutex.Lock()
+	defer d.readyMutex.Unlock()
+	d.ready = true
+}
+
+func (d *waterImpl) close() {
+	close(d.finishedToLoad)
+}
+
 func (d *waterImpl) dialServer(op *ops.Op, ctx context.Context) (net.Conn, error) {
 	return d.reportDialCore(op, func() (net.Conn, error) {
-		d.wgDownload.Wait()
-		if d.dialer == nil {
-			return nil, log.Errorf("dialer not available: %w", d.downloadErr)
+		// if dialer is not ready, wait until WASM is downloaded or context timeout
+		ready, err := d.isReady()
+		if err != nil {
+			return nil, log.Errorf("failed to load WASM for dialer: %w", err)
+		}
+
+		if !ready {
+			select {
+			case _, ok := <-d.finishedToLoad:
+				if !ok {
+					d.errLoadingWASM = log.Error("Channel closed")
+					return nil, d.errLoadingWASM
```

Comment:
This error is kind of mysterious to callers of `dialServer` since they don't know anything about a channel or why it would be closed.  The channel should only be closed if `waterImpl` is closed, right?  If that is correct, then maybe this error should read "dialer closed".

Reply:
It also looks like there is unprotected writing of `d.errLoadingWASM` happening here.  `dialServer` could be called concurrently by multiple goroutines, right?

Reply:
> It also looks like there is unprotected writing of d.errLoadingWASM happening here. dialServer could be called concurrently by multiple goroutines, right?

You're right, that validation should only happen when it receives the signal that it finished to load. Thanks for pointing that!

Reply:
Wait, nop, this verification is only happening after receiving a `ready` dialer. We probably don't need this verification